### ISSUE TITLE
feat: add `file-transfer` plugin

### ISSUE THREAD

**Issue by robingenz** [capawesome-team/capacitor-plugins#24]
**Is your feature request related to a problem? Please describe:**

Currently there is no Capacitor 4 compatible plugin that can be used to upload and download files.

**Describe the solution you'd like:**

Create a plugin that allows you to upload and download files.

**Describe alternatives you've considered:**

**Additional context:**

Related to https://github.com/ionic-team/capacitor/issues/5964

**Comment by Tragio** [comment]
@robingenz there is this plugin https://github.com/AlwaysLoveme/capacitor-plugin-filedownload not sure if can help or contribute to it.

**Comment by muuvmuuv** [comment]
Capacitor Filesystem will soon be able to download files and Http has been updated to handle FormData and files in FormData.

**Comment by robingenz** [comment]
This plugin will offer advanced features. That is why it is still on the roadmap, at least for now.

**Comment by muuvmuuv** [comment]
Cool! Then I will add my "request" for an advanced-feature: progress event callback for each download instead/(additionally) of a listener on all request (like it is with the current implementation by Ionic-Team). 

Our use-case is a download-widget which downloads >1800 files in parallel in batches and I would like to have a proper kb/s and estimated time progress in it. Our current implementation works but it requires to cache the url and loaded bytes in a Set and other hacky stuff to make it work.

**Comment by muuvmuuv** [comment]
Ah, great, that would work a little better if it is returned with each event. 

Just sharing my listener here for anyone looking for that:

<details>
<summary>Http download listener</summary>

```ts
private createHttpListener() {
  return Filesystem.addListener("progress", ({ url, bytes, contentLength }) => {
    const progress = new DownloadProgress(
      this.remaining,
      this.bytesTotal,
      this.bytesLoaded,
      contentLength,
    )

    if (!url.includes("your-domain.")) {
      // Files coming from other servers have no file size so we need to get that
      // value during file download.
      if (contentLength === 0) {
        // Some servers do not respond with a content length.
        this.event.downloadProgressChange.next(progress)
        return
      }
      if (!this.previousBytes.has(url)) {
        this.bytesTotal += contentLength
        this.previousBytes.set(url, 0)
      }
    }

    this.bytesLoaded += bytes - (this.previousBytes.get(url) || 0)
    this.previousBytes.set(url, bytes)

    this.event.downloadProgressChange.next(progress)
  })
}
```

</details>

**Comment by SchneMa** [comment]
Are there any infos when or in which version Upload/Download will be available?

**Comment by robingenz** [comment]
@SchneMa ETA is next month.

**Comment by coderpradp** [comment]
Any news on this?

**Comment by robingenz** [comment]
> Any news on this?

I prioritized the development of another plugin for a sponsor (see https://github.com/capawesome-team/capacitor-firebase/pull/434). Therefore, the development of this plugin has been postponed. I plan to finish it this month.

**Comment by lsolano2707** [comment]
Any news?, I want to upload large files (videos) with capacitor without Ionic and I read "The File Handling Guide for Capacitor" and here mentions that there are plugins to upload files directly to server but I have no idea what they are

**Comment by robingenz** [comment]
@lsolano2707 A few plugins are:
- [Firebase Cloud Storage](https://capawesome.io/plugins/firebase/cloud-storage/)
- [Cloudinary](https://capawesome.io/plugins/cloudinary/)

But this plugin is not yet available.

**Comment by robingenz** [comment]
Now available: https://capawesome.io/docs/sdks/capacitor/file-transfer/
